### PR TITLE
Added a base64 encoder/decoder for DAGTransaction.

### PR DESCRIPTION
**Pull request by cngo-github**
__PR Category__: Feature

Added a base64 encoder/decoder for DAGTransaction.

### CODE REVIEW

File: modules/rosetta/src/main/scala/org/tessellation/rosetta/domain/EncodedRosettaTransaction.scala
Comment (marcinwadon) [inline modules/rosetta/src/main/scala/org/tessellation/rosetta/domain/EncodedRosettaTransaction.scala:18]: you can prepare `Base64Codec` in the same way as the `BinaryCodec` is created. it doesn't need to be transaction-specific

**Review by marcinwadon — CHANGES_REQUESTED** [review]

File: modules/rosetta/src/main/scala/org/tessellation/rosetta/domain/EncodedRosettaTransaction.scala
Comment (cngo-github) [reply modules/rosetta/src/main/scala/org/tessellation/rosetta/domain/EncodedRosettaTransaction.scala:18]: Is this what you are referring to?

https://github.com/Constellation-Labs/tessellation/blob/37466d8b50bbb69a4010df3b293c0664bee9fb4f/modules/sdk/src/main/scala/org/tessellation/sdk/domain/healthcheck/consensus/types/ConsensusHealthStatus.scala#L19-L25

Comment (marcinwadon) [reply modules/rosetta/src/main/scala/org/tessellation/rosetta/domain/EncodedRosettaTransaction.scala:18]: No, this is the usage of the BinaryCodec. Please take a look into the implementation

File: modules/shared/src/main/scala/org/tessellation/ext/codecs.scala
Comment (marcinwadon) [inline modules/shared/src/main/scala/org/tessellation/ext/codecs.scala:49]: Is Base64 an effect type? I don't think so 🤔

File: modules/shared/src/main/scala/org/tessellation/ext/codecs.scala
Comment (marcinwadon) [inline modules/shared/src/main/scala/org/tessellation/ext/codecs.scala:51]: why new String is required? It looks like a plain java :D

Comment (cngo-github) [reply modules/shared/src/main/scala/org/tessellation/ext/codecs.scala]: Changed to match PR https://github.com/Constellation-Labs/tessellation/pull/592

File: modules/shared/src/main/scala/org/tessellation/ext/codecs.scala
Comment (cngo-github) [reply modules/shared/src/main/scala/org/tessellation/ext/codecs.scala]: Changed to match PR https://github.com/Constellation-Labs/tessellation/pull/592

File: modules/rosetta/src/main/scala/org/tessellation/rosetta/domain/EncodedRosettaTransaction.scala
Comment (marcinwadon) [inline modules/rosetta/src/main/scala/org/tessellation/rosetta/domain/EncodedRosettaTransaction.scala:13]: Isn't it just Hex?

File: modules/rosetta/src/test/scala/org/tessellation/rosetta/domain/construction/ConstructionServiceSuite.scala
Comment (marcinwadon) [inline modules/rosetta/src/test/scala/org/tessellation/rosetta/domain/construction/ConstructionServiceSuite.scala:314]: Leftover

File: modules/rosetta/src/test/scala/org/tessellation/rosetta/domain/construction/ConstructionServiceSuite.scala
Comment (marcinwadon) [inline modules/rosetta/src/test/scala/org/tessellation/rosetta/domain/construction/ConstructionServiceSuite.scala:315]: It will always succeed

File: modules/rosetta/src/test/scala/org/tessellation/rosetta/domain/construction/ConstructionServiceSuite.scala
Comment (cngo-github) [reply modules/rosetta/src/test/scala/org/tessellation/rosetta/domain/construction/ConstructionServiceSuite.scala:314]: This test was to test the decoder and encoded. I can still not find the implicits for encoding and decoding `Hashed[Transaction]]` here:

https://github.com/cngo-github/tessellation/blob/20230307-base64/modules/rosetta/src/test/scala/org/tessellation/rosetta/domain/construction/ConstructionServiceSuite.scala#L311

Why?

File: modules/rosetta/src/test/scala/org/tessellation/rosetta/domain/construction/ConstructionServiceSuite.scala
Comment (cngo-github) [reply modules/rosetta/src/test/scala/org/tessellation/rosetta/domain/construction/ConstructionServiceSuite.scala:315]: This test was to test the decoder and encoded. I can still not find the implicits for encoding and decoding `Hashed[Transaction]]` here:

https://github.com/cngo-github/tessellation/blob/20230307-base64/modules/rosetta/src/test/scala/org/tessellation/rosetta/domain/construction/ConstructionServiceSuite.scala#L311

Why?

File: modules/rosetta/src/main/scala/org/tessellation/rosetta/domain/EncodedRosettaTransaction.scala
Comment (cngo-github) [reply modules/rosetta/src/main/scala/org/tessellation/rosetta/domain/EncodedRosettaTransaction.scala:13]: Changed.

**Comment by cngo-github** [comment]
Re-implemented `DAGTransaction` and `Signed[DAGTransaction]` using the `JsonSerializer` from PR https://github.com/Constellation-Labs/tessellation/pull/592.

File: modules/rosetta/src/main/scala/org/tessellation/rosetta/domain/construction/ConstructionService.scala
Comment (marcinwadon) [inline modules/rosetta/src/main/scala/org/tessellation/rosetta/domain/construction/ConstructionService.scala:111]: you need to cover now the newly added function (see below)

File: modules/rosetta/src/test/scala/org/tessellation/rosetta/domain/construction/ConstructionServiceSuite.scala
Comment (marcinwadon) [inline modules/rosetta/src/test/scala/org/tessellation/rosetta/domain/construction/ConstructionServiceSuite.scala:44]: what's the purpose of using tagless algebra here? also, import `schema.transaction.DAGTransaction` should be fixed

File: modules/rosetta/src/test/scala/org/tessellation/rosetta/domain/construction/ConstructionServiceSuite.scala
Comment (marcinwadon) [reply modules/rosetta/src/test/scala/org/tessellation/rosetta/domain/construction/ConstructionServiceSuite.scala:47]: `JsonSerializer.serialize` should return an effect type, otherwise it's a pure function. (this holds for the whole codebase, except interoperability with java or some external impure libraries). what's more - JsonSerializer has been added in this pull request, so you should know/check how it behaves. if it's not pure, then it should be fixed directly in the `JsonSerializer` not to publish impure interface outside.

File: modules/rosetta/src/test/scala/org/tessellation/rosetta/domain/construction/ConstructionServiceSuite.scala
Comment (marcinwadon) [reply modules/rosetta/src/test/scala/org/tessellation/rosetta/domain/construction/ConstructionServiceSuite.scala:44]: imo it's unnecessary overhead. this serialization is just a pure function, so it can be just a simple function, nothing extraordinary. also, it's overhead to use tagless algebra in tests, where everything is using IO here.
imo, there is no point for using for comprehension for a single flat map, but that's up to you.

```scala
generateTransactions(srcAddress, srcKey, dstAddress, txCount).flatMap {
  _.traverse(hashedTransaction =>
    val bytes = getBytes(hashedTransaction, wantSignedTransaction)
            
    (Hex.fromBytes(bytes), hashedTransaction.hash, hashedTransaction)
}
```

with for comp, it looks like that:

```scala
for {
  transactions <- generateTransactions(srcAddress, srcKey, dstAddress, txCount)
  testValues <- transactions.traverse { hashedTransaction =>
    val bytes = getBytes(hashedTransaction, wantSignedTransaction)
    
    (Hex.fromBytes(bytes), hashedTransaction.hash, hashedTransaction)
  }
} yield testValues
```

File: modules/rosetta/src/test/scala/org/tessellation/rosetta/domain/construction/ConstructionServiceSuite.scala
Comment (sdavidp) [reply modules/rosetta/src/test/scala/org/tessellation/rosetta/domain/construction/ConstructionServiceSuite.scala:47]:
```
  def serialize[A: Encoder](content: A) =
    content.asJson.noSpaces.getBytes("UTF-8")
```
I had trouble bring the codebase down but I can see this definition, which has no return type. When I duplicate this in the IDE it returns `Array[Byte]`. When I follow the Circe code there is a place where they specifically say the one of the calls they use is unsafe.

So I agree, this (and the decode?) should return its result in an effect.

Comment (marcinwadon) [reply modules/rosetta/src/test/scala/org/tessellation/rosetta/domain/construction/ConstructionServiceSuite.scala:47]: and there should be a return type. I think I've already commented that in the original PR

Comment (sdavidp) [reply modules/rosetta/src/test/scala/org/tessellation/rosetta/domain/construction/ConstructionServiceSuite.scala:47]: > btw. can you point me to the documentation/code which is unsafe? I cannot find it.

I'm having trouble retracing my steps :( I'll keep looking.
However, `getBytes` is a java function and it throws, so if we want to make `serialize` safer we need to something...

Comment (marcinwadon) [reply modules/rosetta/src/test/scala/org/tessellation/rosetta/domain/construction/ConstructionServiceSuite.scala:47]: :+1: I agree. @kriscieslak can you take a look?

File: modules/rosetta/src/test/scala/org/tessellation/rosetta/domain/construction/ConstructionServiceSuite.scala
Comment (cngo-github) [reply modules/rosetta/src/test/scala/org/tessellation/rosetta/domain/construction/ConstructionServiceSuite.scala:44]: Fixed the import.

Comment (cngo-github) [reply modules/rosetta/src/test/scala/org/tessellation/rosetta/domain/construction/ConstructionServiceSuite.scala]: The serializer was made safe in PR https://github.com/Constellation-Labs/tessellation/pull/607

Comment (marcinwadon) [reply modules/rosetta/src/test/scala/org/tessellation/rosetta/domain/construction/ConstructionServiceSuite.scala]: Yep. Are you going to rebase against that branch?

Comment (cngo-github) [reply modules/rosetta/src/test/scala/org/tessellation/rosetta/domain/construction/ConstructionServiceSuite.scala]: Removed the tagless algebra from the tests.

File: modules/rosetta/src/test/scala/org/tessellation/rosetta/domain/construction/ConstructionServiceSuite.scala
Comment (cngo-github) [reply modules/rosetta/src/test/scala/org/tessellation/rosetta/domain/construction/ConstructionServiceSuite.scala:44]: The delay has been removed.

File: modules/rosetta/src/test/scala/org/tessellation/rosetta/domain/construction/ConstructionServiceSuite.scala
Comment (cngo-github) [reply modules/rosetta/src/test/scala/org/tessellation/rosetta/domain/construction/ConstructionServiceSuite.scala]: @marcinwadon , I was going to wait until that branch gets merged to develop.

File: modules/rosetta/src/main/scala/org/tessellation/rosetta/domain/construction/ConstructionService.scala
Comment (cngo-github) [reply modules/rosetta/src/main/scala/org/tessellation/rosetta/domain/construction/ConstructionService.scala:111]: See comments below.

**Comment by cngo-github** [comment]
This PR is superseded by PR https://github.com/Constellation-Labs/tessellation/pull/627.

**Pull request closed without merging**
